Regression: Fail to evaluate static slice with reference to internally-mutable data

<!--
Thank you for filing a regression report! 🐛 A regression is something that changed between versions of Rust but was not supposed to.

Please provide a short summary of the regression, along with any information you feel is relevant to replicate it.
-->

### Code

I tried this code:

```rust
use std::sync::OnceLock;

static LAZY_INIT: OnceLock<u32> = OnceLock::new();

static LAZY_INIT_REF: &[&OnceLock<u32>] = &[&LAZY_INIT];
```

I expected to see this happen: it compiles with `LAZY_INIT` being reachable through `LAZY_INIT_REF`.

Instead, this happened:

```
error[E0080]: it is undefined behavior to use this value
 --> src/main.rs:5:1
  |
5 | static LAZY_INIT_REF: &[&OnceLock<u32>] = &[&LAZY_INIT];
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ constructing invalid value at .<deref>[0]: encountered reference to mutable memory in `const`
  |
  = note: The rules on what exactly is undefined behavior aren't clear, so this check might be overzealous. Please open an issue on the rustc repository if you believe it should not be considered undefined behavior.
  = note: the raw bytes of the constant (size: 8, align: 8) {
              ╾─────alloc4<imm>─────╼                         │ ╾──────╼
          }

note: erroneous constant encountered
 --> src/main.rs:5:43
  |
5 | static LAZY_INIT_REF: &[&OnceLock<u32>] = &[&LAZY_INIT];
  |                                           ^^^^^^^^^^^^^
```

### Version it worked on

<!--
Provide the most recent version this worked on, for example:

It most recently worked on: Rust 1.47
-->

It most recently worked on: `nightly-2024-02-10`

### Version with regression

`nightly-2024-02-11`

`rustc --version --verbose`:
```
rustc 1.78.0-nightly (6cc484351 2024-02-10)
binary: rustc
commit-hash: 6cc4843512d613f51ec81aba689180c31b0b28b6
commit-date: 2024-02-10
host: aarch64-apple-darwin
release: 1.78.0-nightly
LLVM version: 17.0.6
```

@rustbot modify labels: +regression-from-stable-to-nightly +A-const-eval +T-compiler -regression-untriaged

--- a/tests/ui/consts/static-promoted-to-mutable-static.rs
+++ b/tests/ui/consts/static-promoted-to-mutable-static.rs
@@ -1,0 +1,33 @@
+// check-pass
+#![allow(non_camel_case_types, non_upper_case_globals, static_mut_ref)]
+
+pub struct wl_interface {
+    pub version: i32
+}
+
+pub struct Interface {
+    pub other_interfaces: &'static [&'static Interface],
+    pub c_ptr: Option<&'static wl_interface>,
+}
+
+pub static mut wl_callback_interface: wl_interface = wl_interface {
+    version: 0,
+};
+
+pub static WL_CALLBACK_INTERFACE: Interface = Interface {
+    other_interfaces: &[],
+    c_ptr: Some(unsafe { &wl_callback_interface }),
+};
+
+// This static contains a promoted that points to a static that points to a mutable static.
+pub static WL_SURFACE_INTERFACE: Interface = Interface {
+    other_interfaces: &[&WL_CALLBACK_INTERFACE],
+    c_ptr: None,
+};
+
+// And another variant of the same thing, this time with interior mutability.
+use std::sync::OnceLock;
+static LAZY_INIT: OnceLock<u32> = OnceLock::new();
+static LAZY_INIT_REF: &[&OnceLock<u32>] = &[&LAZY_INIT];
+
+fn main() {}

